### PR TITLE
gh-104683: Argument Clinic: Refactor and simplify 'add docstring' states

### PR DESCRIPTION
**Pull request by erlend-aasland**
Introduce docstring_append() helper, and use it for both parameter and
function docstrings. Remove docstring fixup from
do_post_block_processing_cleanup(); instead, make sure no fixup is needed.

* Issue: gh-104683

### CODE REVIEW

File: Tools/clinic/clinic.py [hunk 1]
```diff
@@ -4617,15 +4617,18 @@ def parse(self, block: Block) -> None:
                 fail("'preserve' only works for blocks that don't produce any output!")
             block.output = self.saved_output
 
-    @staticmethod
-    def valid_line(line: str) -> bool:
+    def in_docstring(self) -> bool:
+        """Return true if we are processing a docstring."""
+        return "docstring" in self.state.__name__
```
Comment (erlend-aasland) [inline Tools/clinic/clinic.py]: Perhaps this is a tad too hacky.

Comment (erlend-aasland) [reply Tools/clinic/clinic.py]:
```suggestion
        return self.state in {
            self.state_parameter_docstring,
            self.state_function_docstring,
        }
```

Comment (AlexWaygood) [reply Tools/clinic/clinic.py]: > Perhaps this is a tad too hacky.

uhhh yes :)

The current PR's implementation of `in_docstring()` returns `True` if `self.state` is `state_parameter_docstring`, `state_function_docstring` or `state_parameter_docstring_start`. Your suggested change only accounts for two of those states. Which is correct? :)

File: Tools/clinic/clinic.py [hunk 2]
```diff
@@ -5281,16 +5292,8 @@ def state_parameter_docstring(self, line: str) -> None:
...
-
-        new_docstring = last_parameter.docstring
-
-        if new_docstring:
-            new_docstring += '\n'
-        if stripped:
-            new_docstring += self.indent.dedent(line)
-
-        last_parameter.docstring = new_docstring
+        last_param = list(self.function.parameters.values())[-1]
```
Comment (AlexWaygood) [inline Tools/clinic/clinic.py]: I think it's more idiomatic to do as the existing code does, and call `next(reversed(it))` to get the last value of an iterable. But we can drop the unnecessary cast to `list()` and iterate directly over the values:

```suggestion
        last_param = next(reversed(self.function.parameters.values()))
```

Comment (erlend-aasland) [reply Tools/clinic/clinic.py]: Yeah, `next(reversed(it))` is fine

File: Tools/clinic/clinic.py (hunk 1 above)
Comment (erlend-aasland) [reply Tools/clinic/clinic.py]: Yes, we only want this in the parsing states, not the docstring-start state :)

File: Tools/clinic/clinic.py [hunk 3]
```diff
@@ -5580,12 +5577,6 @@ def do_post_block_processing_cleanup(self) -> None:
             if no_parameter_after_star:
                 fail("Function " + self.function.name + " specifies '*' without any parameters afterwards.")
 
-        # remove trailing whitespace from all parameter docstrings
-        for name, value in self.function.parameters.items():
-            if not value:
-                continue
-            value.docstring = value.docstring.rstrip()
```
Comment (AlexWaygood) [inline Tools/clinic/clinic.py:5587]: Do we have a test that ensures the parameter docstrings have no trailing whitespace? Seems like no tests fail if I apply just this diff to `main` -- was it just always redundant?

```diff
--- a/Tools/clinic/clinic.py
+++ b/Tools/clinic/clinic.py
@@ -5580,12 +5580,6 @@ def do_post_block_processing_cleanup(self) -> None:
             if no_parameter_after_star:
                 fail("Function " + self.function.name + " specifies '*' without any parameters afterwards.")

-        # remove trailing whitespace from all parameter docstrings
-        for name, value in self.function.parameters.items():
-            if not value:
-                continue
-            value.docstring = value.docstring.rstrip()
-
         self.function.docstring = self.format_docstring()
```

Comment (AlexWaygood) [reply Tools/clinic/clinic.py:5587]: > was it just always redundant?

No, it was just untested, not redundant: a test _does_ fail if I apply _this_ diff to `main`:

```diff
--- a/Tools/clinic/clinic.py
+++ b/Tools/clinic/clinic.py
@@ -5584,7 +5584,7 @@ def do_post_block_processing_cleanup(self) -> None:
         for name, value in self.function.parameters.items():
             if not value:
                 continue
-            value.docstring = value.docstring.rstrip()
+            assert value.docstring == value.docstring.rstrip()

         self.function.docstring = self.format_docstring()
```

(`ClinicParserTest.test_function_docstring`)

**Review by AlexWaygood — CHANGES_REQUESTED** [review]
> Remove docstring fixup from
do_post_block_processing_cleanup(); instead, make sure no fixup is needed.

I think the fixup is still needed. If I apply this diff to your PR branch:

```diff
--- a/Tools/clinic/clinic.py
+++ b/Tools/clinic/clinic.py
@@ -5577,6 +5577,11 @@ def do_post_block_processing_cleanup(self) -> None:
             if no_parameter_after_star:
                 fail("Function " + self.function.name + " specifies '*' without any parameters afterwards.")

+        for name, value in self.function.parameters.items():
+            if not value:
+                continue
+            assert value.docstring == value.docstring.rstrip()
+
         self.function.docstring = self.format_docstring()
```

Then the assertion fails in `ClinicParserTest.test_function_docstring`:

```pytb
======================================================================
FAIL: test_function_docstring (test.test_clinic.ClinicParserTest.test_function_docstring)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\alexw\coding\cpython\Lib\test\test_clinic.py", line 632, in test_function_docstring
    function = self.parse_function("""
               ^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\alexw\coding\cpython\Lib\test\test_clinic.py", line 1384, in parse_function
    block = self.parse(text)
            ^^^^^^^^^^^^^^^^
  File "C:\Users\alexw\coding\cpython\Lib\test\test_clinic.py", line 1380, in parse
    parser.parse(block)
  File "C:\Users\alexw\coding\cpython\Tools\clinic\clinic.py", line 4612, in parse
    self.do_post_block_processing_cleanup()
  File "C:\Users\alexw\coding\cpython\Tools\clinic\clinic.py", line 5583, in do_post_block_processing_cleanup
    assert value.docstring == value.docstring.rstrip()
AssertionError

----------------------------------------------------------------------
```

File: Tools/clinic/clinic.py (hunk 3 above)
Comment (erlend-aasland) [reply Tools/clinic/clinic.py:5587]: I'll add a test :)

**Comment by erlend-aasland** [comment]
> I think the fixup is still needed. [...]

The docstrings are further processed in `format_docstring`; I'm not sure it matters that they are off at this point in the parsing.

**Comment by AlexWaygood** [comment]
> > I think the fixup is still needed. [...]
> 
> The docstrings are further processed in `format_docstring`; I'm not sure it matters that they are off at this point in the parsing.

Ugh, yes, I think you're right :)

**Review by AlexWaygood — APPROVED** [review]
Thank you! Looks good now the test's been added :)

**Comment by erlend-aasland** [comment]
> Ugh, yes, I think you're right :)

See also the added test ;)

**Pull request merged**